### PR TITLE
fix peerstore in hostv2

### PR DESCRIPTION
**Pull request by LeoHChen**
add support of proper PeerID in hostv2 for libp2p support. The PeerID is generated from the public key of the key pair. It will be used in secure communication among nodes.
we were using insecure mode in libp2p before.  this changeset adds the proper key pair and the right peerID support.

It contains many changes as we have to propagate the PeerID in the p2p.peer structure and every node needs to get the other party's PeerID in order to talk to them.

### CODE REVIEW

File: cmd/client/txgen/main.go [hunk 1]
```diff
@@ -59,19 +63,35 @@ func main() {
 	// Add GOMAXPROCS to achieve max performance.
 	runtime.GOMAXPROCS(1024)
 
-	var clientPeer *p2p.Peer
+	var BCPeer *p2p.Peer
```
Comment (rlan35) [inline cmd/client/txgen/main.go]: Better be "bcPeer"?

File: p2p/host/hostv2/hostv2.go [hunk 2]
```diff
@@ -24,8 +24,38 @@ const (
...
+	targetAddr, err := multiaddr.NewMultiaddr(addr)
+	if err != nil {
+		log.Error("AddPeer NewMultiaddr error", "error", err)
+		return err
+	}
+
+	p.Addrs = append(p.Addrs, targetAddr)
+
+	host.Peerstore().AddAddrs(p.PeerID, p.Addrs, peerstore.PermanentAddrTTL)
+	fmt.Printf("AddPeer add to peerstore: %v\n", *p)
```
Comment (rlan35) [inline p2p/host/hostv2/hostv2.go]: Use log.Info

**Comment by mikedoan** [comment]
Sorry. I have not finished the review yet.

File: p2p/p2pimpl/p2pimpl.go [hunk 3]
```diff
@@ -1,25 +1,57 @@
...
+	"net"
+
 	"github.com/harmony-one/harmony/p2p"
 	"github.com/harmony-one/harmony/p2p/host"
 	"github.com/harmony-one/harmony/p2p/host/hostv1"
 	"github.com/harmony-one/harmony/p2p/host/hostv2"
+
+	"github.com/harmony-one/harmony/internal/utils"
+	peer "github.com/libp2p/go-libp2p-peer"
+	multiaddr "github.com/multiformats/go-multiaddr"
```
Comment (harmony-ek) [inline p2p/p2pimpl/p2pimpl.go]: [nit/style] This package is `ma` by default, and people's eyes have been trained to expect `ma` to mean the multiaddr package.  Do we have a strong reason to use an alternative package tag?

File: p2p/p2pimpl/p2pimpl.go [hunk 4]
```diff
@@ -1,25 +1,57 @@
...
+// on the same host if the node software restarted. The peerstore has to be persistent as well.
+func NewHost(self *p2p.Peer) (host.Host, error) {
 	if Version == 1 {
-		h := hostv1.New(peer)
-		return h
+		h := hostv1.New(self)
+		return h, nil
+	}
+
+	selfAddr, err := multiaddr.NewMultiaddr(fmt.Sprintf("/ip4/0.0.0.0/tcp/%s", self.Port))
```
Comment (harmony-ek) [inline p2p/p2pimpl/p2pimpl.go]: Just a note for future that the port may be an ephemeral port later when we have node discovery, which is a node ID → node locator (such as a multiaddr) lookup service.  In that case, a node would just let the OS pick an unused port (in the socket interface one does this by saying port 0; unsure what the libp2p way is) and bind to it, then later register its node-to-locator association with the node discovery/rendezvous service, using its own IP address + the ephemeral port number.

(This logic gets further complicated when we consider NAT traversal, but let's not go too deep :-p)

File: p2p/p2pimpl/p2pimpl.go [hunk 5]
```diff
@@ -1,25 +1,57 @@
...
+	if err != nil {
+		return nil, err
+	}
+	self.Addrs = append(self.Addrs, selfAddr)
+
+	// TODO (leo), change to GenKeyP2PRand() to generate random key. Right now, the key is predicable as the
```
Comment (harmony-ek) [inline p2p/p2pimpl/p2pimpl.go]:
```suggestion
	// TODO (leo), change to GenKeyP2PRand() to generate random key. Right now, the key is predictable as the
```

File: consensus/consensus_leader_msg_test.go [hunk 6]
```diff
@@ -14,7 +14,7 @@ import (
 func TestConstructAnnounceMessage(test *testing.T) {
 	leader := p2p.Peer{IP: "127.0.0.1", Port: "19999"}
 	validator := p2p.Peer{IP: "127.0.0.1", Port: "55555"}
-	host := p2pimpl.NewHost(leader)
+	host, _ := p2pimpl.NewHost(&leader)
```
Comment (harmony-ek) [inline consensus/consensus_leader_msg_test.go]: Why ignore error?
```suggestion
	host, err := p2pimpl.NewHost(&leader)
	if err != nil {
		test.FailNow(fmt.Sprintf("cannot create host: %s", err))
	}
```

File: cmd/beaconchain/main.go [hunk 7]
```diff
@@ -48,6 +48,9 @@ func main() {
 		beaconchain.SetSaveFile(*resetFlag)
 		bc = beaconchain.New(*numShards, *ip, *port)
 	}
+
+	fmt.Printf("Beacon Chain Started: /ip4/%s/tcp/%v/ipfs/%s\n", *ip, *port, bc.GetID().Pretty())
```
Comment (mikedoan) [inline cmd/beaconchain/main.go]: Just some thinking. do you think we need to use log for all the printing and make it consistently over the whole code base?

File: consensus/consensus_leader_test.go [hunk 8]
```diff
@@ -81,17 +84,20 @@ func TestProcessMessageLeaderResponse(test *testing.T) {
...
+	host2, _ := p2pimpl.NewHost(&validator2)
+	consensusValidator2 := New(host2, "0", []p2p.Peer{validator1, validator2, validator3}, leader)
 	consensusValidator2.blockHash = [32]byte{}
 	_, msg = consensusValidator2.constructCommitMessage(consensus_proto.MessageType_COMMIT)
 	consensusLeader.ProcessMessageLeader(msg[1:])
 
-	consensusValidator3 := New(p2pimpl.NewHost(validator3), "0", []p2p.Peer{validator1, validator2, validator3}, leader)
+	host3, _ := p2pimpl.NewHost(&validator3)
+	consensusValidator3 := New(host3, "0", []p2p.Peer{validator1, validator2, validator3}, leader)
 	consensusValidator3.blockHash = [32]byte{}
```
Comment (harmony-ek) [inline consensus/consensus_leader_test.go]: [nit/refactor] Can you roll the above section into a loop to reduce copy-and-paste and make it a bit more readable?

File: cmd/harmony.go [hunk 9]
```diff
@@ -142,7 +167,16 @@ func main() {
 		ldb, _ = InitLDBDatabase(*ip, *port)
 	}
 
-	host := p2pimpl.NewHost(selfPeer)
+	host, err := p2pimpl.NewHost(&selfPeer)
+	if err != nil {
+		panic("unable to new host in harmony")
+	}
+
+	fmt.Printf("[Harmony] selfPeer: %v\n", selfPeer)
```
Comment (mikedoan) [inline cmd/harmony.go]: is it debug message? log.Debug could be better?

File: internal/newnode/newnode.go [hunk 10]
```diff
@@ -72,32 +80,34 @@ func (node *NewNode) requestBeaconChain(BCPeer p2p.Peer) (err error) {
...
 	}
-	p := p2p.Peer{IP: node.Self.IP, Port: node.Self.Port}
-	nodeInfo := &bcconn.NodeInfo{Self: p, PubK: pubk}
+	nodeInfo := &proto_node.Info{IP: node.Self.IP, Port: node.Self.Port, PubKey: pubk, PeerID: node.host.GetID()}
 	msg := bcconn.SerializeNodeInfo(nodeInfo)
 	msgToSend := proto_identity.ConstructIdentityMessage(proto_identity.Register, msg)
 	gotShardInfo := false
-	timeout := time.After(300 * time.Second)
-	tick := time.Tick(5 * time.Second)
+	timeout := time.After(120 * time.Second)
```
Comment (harmony-ek) [inline internal/newnode/newnode.go]: Maybe add a TODO saying that this should be configurable?  Also, maybe `2 * time.Minute` instead?

File: internal/newnode/newnode.go [hunk 11]
```diff
@@ -72,32 +80,34 @@ func (node *NewNode) requestBeaconChain(BCPeer p2p.Peer) (err error) {
...
 				host.SendMessage(node.host, BCPeer, msgToSend, nil)
 			}
 		}
 	}
 	if !gotShardInfo {
 		err = errors.New("could not create connection")
-		node.log.Crit("Could not get sharding info after 5 minutes")
-		os.Exit(1)
+		node.log.Crit("Could not get sharding info after 2 minutes")
+		os.Exit(10)
```
Comment (mikedoan) [inline internal/newnode/newnode.go]: can you elaborate this? 10 or 1?

File: p2p/p2pimpl/p2pimpl.go [hunk 12]
```diff
@@ -1,25 +1,57 @@
...
+	peerID, err := peer.IDFromPublicKey(pubKey)
+
+	if err != nil {
+		return nil, err
+	}
+
+	self.PeerID = peerID
+	h := hostv2.New(*self, priKey)
+
+	fmt.Printf("NewHost => self:%s, PeerID: %v\n", net.JoinHostPort(self.IP, self.Port), self.PeerID)
```
Comment (mikedoan) [inline p2p/p2pimpl/p2pimpl.go]: log.Info - we have a logging singleton which can be used across all code base.

File: p2p/p2pimpl/p2pimpl.go (hunk 4 above)
Comment (LeoHChen) [reply p2p/p2pimpl/p2pimpl.go]: @harmony-ek , please open an issue for further note.

File: consensus/consensus_leader_test.go (hunk 8 above)
Comment (LeoHChen) [reply consensus/consensus_leader_test.go]: don't want to change the original test code too much.  I'll open an issue to simplify the test case as a beginner's task. #284

File: cmd/beaconchain/main.go (hunk 7 above)
Comment (LeoHChen) [reply cmd/beaconchain/main.go]: don't want to change this one now.  current beacon chain cmd will be obsolete soon by boot node.

File: internal/newnode/newnode.go (hunk 11 above)
Comment (LeoHChen) [reply internal/newnode/newnode.go]: make it 10 to make it easier to identify the failure reason in return code of the go test.  Again, newnode will be obsolete soon.

File: p2p/p2pimpl/p2pimpl.go (hunk 12 above)
Comment (LeoHChen) [reply p2p/p2pimpl/p2pimpl.go]: fixed.

File: cmd/harmony.go (hunk 9 above)
Comment (LeoHChen) [reply cmd/harmony.go]: debug message, removed.

**Pull request merged**